### PR TITLE
[poco] fix PocoNetSSL and PocoNetSSLWin simultaneous usage under Windows

### PR DESCRIPTION
**Pull request by mathbunnyru**
Specify library name and version:  **lib/1.0**

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by mathbunnyru** [comment]
Should also fix in CCI:
https://github.com/conan-io/conan-center-index/issues/4184
https://github.com/conan-io/conan-center-index/issues/4095

@paulo-coutinho could you please check, that this change fixed poco for you?

**Comment by paulo-coutinho** [comment]
Hi,

Yes. I that posted the solution on both links.

Thanks.

**Comment by paulo-coutinho** [comment]
With this, i only need "enable_net" correct?

**Comment by paulo-coutinho** [comment]
@mathbunnyru is this that im thinking in do after. Thanks for you time.

**Comment by mathbunnyru** [comment]
> With this, i only need "enable_net" correct?

Yes. Or you could try default options as well if you don't need any components which are not turned on by default.

File: recipes/poco/all/conanfile.py
Comment (SSE4) [inline recipes/poco/all/conanfile.py:46]: should it be `self.settings.os == Windows` instead? my doubt that it may break cross-building scenarios (e.g. Windows -> Linux or Linux -> Windows)

File: recipes/poco/all/test_package/conanfile.py
Comment (SSE4) [inline recipes/poco/all/test_package/conanfile.py:19]: nice

File: recipes/poco/all/conanfile.py
Comment (mathbunnyru) [reply recipes/poco/all/conanfile.py:46]: How can I use self in class namespace?

**Comment by mathbunnyru** [comment]
I don't see any error in the builds.
Everything seems as it should be.

**Comment by paulo-coutinho** [comment]
Im thinking the same. I don't see any error to try help.

**Comment by mathbunnyru** [comment]
@dmn-star as you see, commenting out execution of the util doesn't work.

**Comment by dmn-star** [comment]
Too bad. :-) But tt is strange that the _warnings_ are written on std err.

**Comment by dmn-star** [comment]
> Should also fix in CCI:
> #4184
> #4095

@mathbunnyru If you add Fixes #4184  Fixes #4095 the issues will be closed after the PR is merged.
see https://github.com/conan-io/conan-center-index/pull/4117#issue-550531354

**Comment by mathbunnyru** [comment]
> > Should also fix in CCI:
> > #4184
> > #4095
> 
> @mathbunnyru If you add Fixes #4184 Fixes #4095 the issues will be closed after the PR is merged.
> see [#4117 (comment)](https://github.com/conan-io/conan-center-index/pull/4117#issue-550531354)
> 
> PS: Specify library name and version: poco/all

Thanks, I didn't know about this GitHub feature 👍 
In this case, it won't work, because the issues are already closed :)

File: recipes/poco/all/conanfile.py
Comment (mathbunnyru) [reply recipes/poco/all/conanfile.py:46]: @SSE4 could you elaborate on this, please?

Comment (SSE4) [reply recipes/poco/all/conanfile.py:46]: I don't think it's possible. probably, you need to add both components and then process it in `config_options` or somewhere else.
the issue is running on Linux with `self.settings.os == Windows`, `tools.os_info.is_windows` will return False.

**Comment by jgsogo** [comment]
...and getting a really weird error (tomorrow I'll keep investigating). I get an error with package `6affa419bd4c8c4f073e012613e896789c5186e0`, it is built successfully, but then `conan upload` fails:

```
conan upload poco/1.10.1@#40e37c9f4276b82d2e6ad590b2b7b4ab:6affa419bd4c8c4f073e012613e896789c5186e0 -r conan-upstream 
ERROR: Binary package poco/1.10.1:6affa419bd4c8c4f073e012613e896789c5186e0 not found
```

All the packages (before the failure) are built successfully as you can see in [this table](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/BuildSingleReference/3625/poco/1.10.1/summary.json).

**Comment by dmn-star** [comment]
I see lots of "invalid configuration". Or did you build for your custom profile only?
poco_invalid

**Comment by jgsogo** [comment]
Those `INVALID_CONFIGURATION` are the ones being raised from the `configure()` or `validate()` function (and, if any, from `build()` as well).

**Comment by dmn-star** [comment]
I do not see at first look why these configurations are invalid. And these are the same from https://github.com/conan-io/conan-center-index/pull/4205#issuecomment-760126817

**Comment by jgsogo** [comment]
In this `conanfile.py`, those `InvalidConfiguration` are raised from the `configure()` method or from any of the requirement this library depends on. An InvalidConfiguration is not a failure for the CI, it skips the build.

**Comment by dmn-star** [comment]
I have tried to build POCO on my workstation but without success.

@mathbunnyru Can you even compile POCO through on your end?

 **conan create recipes/poco/all poco/1.10.1@ -pr=conan/profiles/win_poco --build**
Exporting package recipe
poco/1.10.1 exports: File 'conandata.yml' found. Exporting it...
poco/1.10.1 exports: Copied 1 '.yml' file: conandata.yml
poco/1.10.1 exports_sources: Copied 1 '.txt' file: CMakeLists.txt
poco/1.10.1 exports_sources: Copied 6 '.patch' files
poco/1.10.1: The stored package has not changed
poco/1.10.1: Exported revision: 01242c1573bd3894a240b8caaf9ae335
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MT
compiler.version=16
os=Windows
os_build=Windows
[options]
poco:shared=False
[build_requires]
[env]

WARN: apr-util/1.6.1: requirement expat/2.2.9 overridden by poco/1.10.1 to expat/2.2.10
WARN: libmysqlclient/8.0.17: requirement openssl/1.1.1i overridden by poco/1.10.1 to openssl/1.1.1h
apr/1.7.0: Forced build from source
bzip2/1.0.8: Forced build from source
expat/2.2.10: Forced build from source
openssl/1.1.1h: Forced build from source
sqlite3/3.33.0: Forced build from source
zlib/1.2.11: Forced build from source
apr-util/1.6.1: Forced build from source
libmysqlclient/8.0.17: Forced build from source
libpq/11.5: Forced build from source
pcre/8.41: Forced build from source
poco/1.10.1: Forced build from source
nasm/2.14: Forced build from source
strawberryperl/5.30.0.1: Forced build from source
**ERROR: nasm/2.14: Error in build_requirements() method, line 28
        self.build_requires.add("strawberryperl/5.30.0.1")
        TypeError: add() missing 1 required positional argument: 'context'**

**Comment by dmn-star** [comment]
confusing :-)

conan install strawberryperl/5.30.0.1@
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Installing package: strawberryperl/5.30.0.1
Requirements
    strawberryperl/5.30.0.1 from 'conan-center' - Cache
Packages
    strawberryperl/5.30.0.1:456f15897172eef340fcbac8a70811f2beb26a93 - Cache

Installing (downloading, building) binaries...
strawberryperl/5.30.0.1: Already installed!
strawberryperl/5.30.0.1: Appending PATH environment variable: C:\.conan\506a9d\1\bin

**Comment by dmn-star** [comment]
conan install nasm/2.14@
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Installing package: nasm/2.14
Requirements
    nasm/2.14 from 'conan-center' - Cache
Packages
    nasm/2.14:456f15897172eef340fcbac8a70811f2beb26a93 - Cache

Installing (downloading, building) binaries...
nasm/2.14: Already installed!

**Comment by dmn-star** [comment]
Now I don't understand anything at all. LOL

**Comment by mathbunnyru** [comment]
```
ERROR: nasm/2.14: Error in build_requirements() method, line 28
self.build_requires.add("strawberryperl/5.30.0.1")
TypeError: add() missing 1 required positional argument: 'context'
```

This indicates that you're using not the latest revision of nasm recipe when you're building from source (`--build`).
Try to run `conan create` with `--update` flag.

And `conan install` works because it doesn't build anything (package is present in binary form in cci).

**Comment by mathbunnyru** [comment]
@dmn-star

I deleted `.conan` folder (to be sure to use the latest version) and ran your command. Everything was fine (but took some time, a lot of packages and sources to compile).

**Comment by dmn-star** [comment]
I should not coding on Sunday. :-) 

conan create recipes/poco/all poco/1.10.1@ -pr=conan/profiles/win_poco --build **--update**


>-- Build files have been written to: D:/dev/temp/github/conan_poco/conan-center--index/recipes/poco/all/test_package/build/d85a4aab099651c589a871b1bb6738ab106b193c
Microsoft (R)-Build-Engine, Version 16.8.3+39993bd9d für .NET Framework
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

>  Checking Build System
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt
  test_net_2.cpp
  test_core.cpp
  test_netssl.cpp
  test_sqlite.cpp
  test_crypto.cpp
  test_jwt.cpp
  test_net.cpp
  test_encodings.cpp
  test_util.cpp
  net.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c58
  9a871b1bb6738ab106b193c\bin\net.exe
  tcrypto.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab09965
  1c589a871b1bb6738ab106b193c\bin\crypto.exe
  core.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c5
  89a871b1bb6738ab106b193c\bin\core.exe
  encodings.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099
  651c589a871b1bb6738ab106b193c\bin\encodings.exe
  net_2.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c
  589a871b1bb6738ab106b193c\bin\net_2.exe
  util.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c5
  89a871b1bb6738ab106b193c\bin\util.exe
  netssl.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651
  c589a871b1bb6738ab106b193c\bin\netssl.exe
  sqlite.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651
  c589a871b1bb6738ab106b193c\bin\sqlite.exe
  jwt.vcxproj -> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c58
  9a871b1bb6738ab106b193c\bin\jwt.exe
  Building Custom Rule D:/dev/temp/github/conan_poco/conan-center-index/recipes/poco/all/test_package/CMakeLists.txt

**Comment by dmn-star** [comment]
I noticed that the Windows Firewall blocks the test 
D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c589a871b1bb6738ab106b193c\bin**\net_2.exe**

Perhaps that's the problem.

**Comment by dmn-star** [comment]
@mathbunnyru Let's test what happens when we move ConanInvalidConfiguration from build into validate. And maybe also some from configure.

def build(self):
    if self.options.enable_data_sqlite:
        if self.options["sqlite3"].threadsafe == 0:
            raise ConanInvalidConfiguration("sqlite3 must be built with threadsafe enabled")

**Comment by mathbunnyru** [comment]
> @mathbunnyru While we are at it anyway. Let's test what happens when we move ConanInvalidConfiguration from build into validate. And maybe also some from configure. :-)
> 
> ```
> def build(self):
>     if self.options.enable_data_sqlite:
>         if self.options["sqlite3"].threadsafe == 0:
>             raise ConanInvalidConfiguration("sqlite3 must be built with threadsafe enabled")
> ```

Moved everything to `validate()`.

**Comment by mathbunnyru** [comment]
> I noticed that the Windows Firewall blocks the test **net_2.exe**
> D:\dev\temp\github\conan_poco\conan-center-index\recipes\poco\all\test_package\build\d85a4aab099651c589a871b1bb6738ab106b193c\bin\net_2.exe
> 
> Perhaps that's the problem.

Thanks, that's a very good assumption. I don't get an error like this on my VM. Maybe our setup is different (meh, `Windows`).

I will wait until the last commit builds with the same error (to make sure everything works relatively fine with validate). Then I will try to simply disable net_2 for a while to check your assumption.

**Comment by mathbunnyru** [comment]
As you see, @dmn-star, the problem is not resolved with removing net_2.

**Comment by dmn-star** [comment]
But **net** was also blocked, sorry. 

>if self.options["poco"].enable_net:
        self.run(os.path.join("bin", "net"), run_environment=True)
        # self.run(os.path.join("bin", "net_2"), run_environment=True)

**Comment by mathbunnyru** [comment]
@dmn-star nothing changed unfortunately.

**Comment by dmn-star** [comment]
I'm running out of ideas. :-)

**Comment by ericLemanissier** [comment]
@jgsogo @danimtb What makes the bot think these windows builds fail ? I get comparable output on my computer, and conan returns success status code.

**Comment by danimtb** [comment]
We are testing this PR in our staging environment to figure out what is going on. Sorry for the inconvenience

**Comment by mathbunnyru** [comment]
> We are testing this PR in our staging environment to figure out what is going on. Sorry for the inconvenience

No worries. That's great 👍

**Comment by danimtb** [comment]
Moving this changes to our staging environment proved that the CI is working fine **without** the changes in the test_package. I Have just pushed the changes in the test_package to make sure this is the cause of the CI failure and report back. Thank you

**Comment by mathbunnyru** [comment]
Investigating together with @danimtb 👍

File: recipes/poco/all/conanfile.py
Comment (dmn-star) [reply recipes/poco/all/conanfile.py:49]: I would try replacing tools.os_info.is_windows with self.settings.os == Windows as suggested by @SSE4.

**Comment by dmn-star** [comment]
I would try replacing tools.os_info.is_windows with self.settings.os == Windows as suggested by @SSE4.

**Comment by mathbunnyru** [comment]
Sorry, I'm on a one week vacation, so I won't be able to respond here for a while.

File: recipes/poco/all/conanfile.py
Comment (danimtb) [reply recipes/poco/all/conanfile.py:49]: I think we can leave this option to True, since it is removed in other OSs different from windows

File: recipes/poco/all/conanfile.py
Comment (danimtb) [inline recipes/poco/all/conanfile.py:48]: Same here, we can leave this to True, since this option is deleted in windows

**Comment by danimtb** [comment]
We need to add the changes of test_package/conanfile.py to avoid the error of the consumer in the CI. The issue with the failed CI status is due to Jenkins blocking the test_package executable due to networking. I'd suggest to comment the lines and add the reason like:

```
        // FIXME: CCI Jenkins blocks the request making the build fail
        // s.sendRequest(request);	
        // HTTPResponse response;	
        // std::istream& rs = s.receiveResponse(response);	
        // StreamCopier::copyStream(rs, std::cout);
```

**Comment by dmn-star** [comment]
@danimtb 
We have already turned off the **net** and **net_2** tests for trial, which unfortunately did not help. Are there any more tests going online here?

https://github.com/conan-io/conan-center-index/pull/4205#issuecomment-766346181
https://github.com/conan-io/conan-center-index/pull/4205/commits/c53e6a7a91b20f4fbc1a5802da40a7707af6fcac

**Comment by danimtb** [comment]
@dmn-star I see. I have tested with the commented lines above on our staging environment and it is working. I will try to push the changes here and see what happens

**Comment by mathbunnyru** [comment]
Thanks @danimtb ! 👍

**Comment by mathbunnyru** [comment]
@SpaceIm @SSE4 @ericLemanissier @danimtb please, approve the PR :)
It's time to get this merge 🎉

**Review by danimtb — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by prince-chrismc** [comment]
@jgsogo @danimtb This PR triggered a build on master and it looks stuck.. it's been about ~10 hrs with [master status 'being built'](https://api.github.com/repos/conan-io/conan-center-index/status/d7e0581908730a3ccbe2ce397be28655b4409341).
Looking at the [last build in this PR](https://api.github.com/repos/conan-io/conan-center-index/statuses/cb0a9e88ee8ba5325e632c98334352f19d7f8bde) it should be <4 hrs for the build

It's a PR from hell 🙃

**Comment by jgsogo** [comment]
OMG 😫   

I've stopped and triggered again the build for `master`. CI restarted tonight and it is something we need to work on, how to recover from that circunstance.

**Comment by jngrb** [comment]
@jgsogo @danimtb I think this should be changed again in order to again allow OpenSSL-enabled SSL support under Windows

see my discussion with @mathbunnyru in my issue #6311 und my suggested solution in #6314

**Comment by jgsogo** [comment]
> @jgsogo @danimtb I think this should be changed again in order to again allow OpenSSL-enabled SSL support under Windows
> 
> see my discussion with @mathbunnyru in my issue #6311 und my suggested solution in #6314

Hi, @jngrb ! This kind of thing will always happen from time to time. I know people are doing their best in this repository, but we are humans sometimes :slightly_smiling_face:   I will add another batch of people to the EAP so your PR can run, and people with much more experience than me will notice about it and review.

Anyway, if you are using these recipes in a production environment (cc @Talkless ) I encourage you to **use recipe revisions** and/or lockfiles. That way, you will be using always the same version of the recipe and you can decide when to update and start using modifyied versions.

**Comment by prince-chrismc** [comment]
> If you are using these recipes in a production environment I encourage you to use recipe revisions and/or lockfiles.

Words to live by!

**Pull request merged**
